File: maxmult.jl
#S takes values in [0,1]
struct S <: Real
    val::Real
    function S(v::Real)
        if !(0 <= v <= 1)
            throw(ArgumentError("value must be between 0 and 1"))
        end
        new(v)
    end
end

#This implements a max-mult semiring structure over S.
Base.:+(a::S, b::S) = S(max(a.val,b.val))
Base.:*(a::S, b::S) = S(a.val * b.val)

File: maxmult.py
class S(float):
    def __init__(self, value):
        if not (0 <= value <= 1):
            raise ValueError("Max-Mult Semiring value must be between 0 and 1")
        super().__init__(value)

    def __add__(self, other):
        return S(max(float(self),float(other)))

    def __mul__(self, other):
        return S(float(self) * float(other))